### PR TITLE
`type` directive for serializer to control type field with json-api adapter

### PR DESCRIPTION
**Pull request by youroff**
This pull-requests implements `type` directive for serializer declaration. Which sets type attribute while using json-api adapter. So it's possible to use ProfileSerializer with User model and set type field to 'profile', like this:
```
class ProfileSerializer < ActiveModel::Serializer
  attributes :name
  type 'profile'
end
```

### CODE REVIEW

**Comment by NullVoxPopuli** [comment]
First off, thanks for the PR!

My only concern is that `type` is often the name of a field used in single table inheritance for storing the class name of an object. 

So, I guess that leaves two options:
 - change the name of the method from type to something else?
 - or write a test to simulate the potential problem (probably just rendering a model with a type field, and see if it still gets included in the resulting hash) (this would be my preference)  If the test passes, you're good, and I'll give ya the :+1:  :-)

**Comment by youroff** [comment]
There is a test which actually made me to change an approach. So now there's only class method test, which sets _test class variable. And this is used on the class level again, so it'll not interfer with type attribute.
Here's that test:
https://github.com/rails-api/active_model_serializers/blob/master/test/serializers/attribute_test.rb#L55

**Comment by youroff** [comment]
But I think it would be way smarter to have setting which class name to use (Serializer or Model) AND type directive to set it manually. If you like this idea, I'll add this feature.

**Comment by NullVoxPopuli** [comment]
oh good! thanks for looking for that. :-)

you have my :+1: 

@beauby seems to be the JsonApi Adapter expert though, so I'll let him make the final call.

**Comment by youroff** [comment]
I added functionality to derive resource type from serializer class, however it broke one test, which I'm not sure is really correct. Depends on standard which I'm not sure in. So the question is how we deal with namespaced models/serializers. Method `underscore` converts `::` into `/`, however expected behavior is make it underscore as well. Opinions?

**Comment by beauby** [comment]
Hi @youroff, you are right, I added the possibility to override the type in a pending PR that got lost somewhere. I am totally up for this PR, I'll review it right away.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 1]
```diff
@@ -71,10 +71,16 @@ def serializable_hash_for_single_resource(options)
         end
 
         def resource_identifier_type_for(serializer)
-          if ActiveModel::Serializer.config.jsonapi_resource_type == :singular
-            serializer.object.class.model_name.singular
+          unless serializer.class._type.nil?
+            serializer.class._type
           else
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Avoid `unless` with an `else` clause. Rewrite as an `if` with the positive clause first (https://github.com/bbatsov/ruby-style-guide#no-else-with-unless)

File: lib/active_model/serializer/adapter/json_api.rb [hunk 2]
```diff
@@ -71,10 +71,16 @@ def serializable_hash_for_single_resource(options)
...
-            serializer.object.class.model_name.plural
+            if ActiveModel::Serializer.config.type_source == :model
+              type = serializer.object.class.name.underscore
+            else
+              type = serializer.class.name.underscore.split('_serializer').first
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: This seems a bit overcomplicated. How about `serializer.class.name.chomp('Serializer').underscore`?

**Comment by youroff** [comment]
Done, but for namespace case we have to decide what to change: test or code. I'm not sure what json-api says about namespaced models.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 2 above)
Comment (youroff) [inline lib/active_model/serializer/adapter/json_api.rb]: Ok, wasn't sure what could serve more elegant here.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 3]
```diff
@@ -71,10 +71,16 @@ def serializable_hash_for_single_resource(options)
         end
 
         def resource_identifier_type_for(serializer)
-          if ActiveModel::Serializer.config.jsonapi_resource_type == :singular
-            serializer.object.class.model_name.singular
+          unless serializer.class._type.nil?
+            serializer.class._type
           else
-            serializer.object.class.model_name.plural
+            if ActiveModel::Serializer.config.type_source == :model
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: I think this should be extracted to another PR. I'm generally not in favor of adding too many config options, and with the `type` method you introduced, there's virtually no use for this one, except for allowing wildly different workflows for users, which, as well, I'm not too keen on.

File: lib/active_model/serializer/configuration.rb [hunk 4]
```diff
@@ -8,6 +8,7 @@ module Configuration
         base.config.array_serializer = ActiveModel::Serializer::ArraySerializer
         base.config.adapter = :attributes
         base.config.jsonapi_resource_type = :plural
+        base.config.type_source = :model
```
Comment (beauby) [inline lib/active_model/serializer/configuration.rb]: See my [other remark](https://github.com/rails-api/active_model_serializers/pull/1213/files#r40739506).

**Comment by beauby** [comment]
Thing is, JSON API [explicitly](http://cantremembertheissue.jsonapi.org/) does not say anything about namespaced models. There are PRs in the pipe for better handling of namespaced models, but my personal feeling is that the default should be to remove the namespace for the `type` attribute. We should probably open an RFC to discuss it though.

**Comment by youroff** [comment]
I agree. So I remove namespace for this time to make tests pass?

File: lib/active_model/serializer/configuration.rb (hunk 4 above)
Comment (youroff) [inline lib/active_model/serializer/configuration.rb]: I'm quite lost in comments... which remark exactly?

**Comment by beauby** [comment]
Namespaces shouldn't be a problem here, are they?

File: lib/active_model/serializer/configuration.rb (hunk 4 above)
Comment (beauby) [inline lib/active_model/serializer/configuration.rb]: The one I linked ;) https://github.com/rails-api/active_model_serializers/pull/1213/files#r40739506

**Comment by youroff** [comment]
They caused one test to fail. Last commit fixes it. If we agree to remove namespaces.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 5]
```diff
@@ -71,10 +71,16 @@ def serializable_hash_for_single_resource(options)
         end
 
         def resource_identifier_type_for(serializer)
-          if ActiveModel::Serializer.config.jsonapi_resource_type == :singular
-            serializer.object.class.model_name.singular
+          if serializer.class._type.nil?
+            if ActiveModel::Serializer.config.type_source == :model
+              type = serializer.object.class.name.underscore
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: We shouldn't modify the behavior here (before it read the type from `object.class.model_name` (which is AR-specific, so we may change it later, but not now) to `object.class.name`).

**Comment by beauby** [comment]
What caused the test to fail was that you got the name from the object's class name instead of the AR model name.

**Comment by youroff** [comment]
Right, but it doesn't really matter. Anyway we should decide either we leave namespace with separation of some form (`/` or `_`) or we get rid of it.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 5 above)
Comment (youroff) [inline lib/active_model/serializer/adapter/json_api.rb]: The problem is that weird conversion of namespace comes from `plural` or `singular` methods. I Thought it would be reasonable to rely on general pluralization applied to string. But I can roll back to previous behavior of course.

**Comment by youroff** [comment]
http://jsonapi.org/format/#document-member-names
At least it says that `/` is not allowed.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 6]
```diff
@@ -71,10 +71,16 @@ def serializable_hash_for_single_resource(options)
         end
 
         def resource_identifier_type_for(serializer)
-          if ActiveModel::Serializer.config.jsonapi_resource_type == :singular
-            serializer.object.class.model_name.singular
+          if serializer.class._type.nil?
+            if ActiveModel::Serializer.config.type_source == :model
+              type = serializer.object.class.name.demodulize.underscore
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: I'd rather we stick with `model_name`, because it might very well be that there are 0 differences, but I have no certainty of that, and would rather change it in a subsequent PR if it makes sense. I get that it makes the pluralizing situation trickier, but it will be refactored soon anyways.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 7]
```diff
@@ -71,10 +71,16 @@ def serializable_hash_for_single_resource(options)
         end
 
         def resource_identifier_type_for(serializer)
-          if ActiveModel::Serializer.config.jsonapi_resource_type == :singular
-            serializer.object.class.model_name.singular
+          if serializer.class._type.nil?
+            if ActiveModel::Serializer.config.type_source == :model
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Let's loose the `config.type_source` for this PR, maybe extract it to another PR, but I personally favor the `type` directive way.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 8]
```diff
@@ -71,10 +71,16 @@ def serializable_hash_for_single_resource(options)
...
+              type = serializer.object.class.name.demodulize.underscore
+            else
+              type = serializer.class.name.demodulize.chomp('Serializer').underscore
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Also, I'm totally in favor of removing the namespace, but I think it should be done in a separate PR.

**Comment by youroff** [comment]
So I roll everything back to `type` directive? I agree to get type from model_name, but why can't we leave type_source? With default setting (:model) it's transparent and `type` will override it anyways (if needed). I mean it's the feature that I would use in my project, so I'm really interested to keep it.

**Comment by beauby** [comment]
It's not that we can't leave `type_source`, it's just that PRs should be kept so one semantic change as much as possible, so that people can review them more easily, and the history remains clean.

**Comment by youroff** [comment]
Btw, what `plural` and `singular` on `ActiveModel::Name` does is exactly the same as we did:
https://github.com/rails/rails/blob/0edb6465971f7d937fce2bf0a8e1e2b540d56e0a/activemodel/lib/active_model/naming.rb#L193-L195

**Comment by youroff** [comment]
Ok, so I'll just split this to two PR's?

**Comment by beauby** [comment]
Does it? https://github.com/rails/rails/blob/0edb6465971f7d937fce2bf0a8e1e2b540d56e0a/activemodel/lib/active_model/naming.rb#L235

**Comment by beauby** [comment]
That would be nice, yes.

**Comment by youroff** [comment]
Well, in general case it'll return ActiveModel::Name populated with class name of the model, but there are some edge cases, that I was not aware of (and I still don't understand the use-case). So probably we need tests for that weird case to make sure that model_name return exactly what we expect.

**Comment by youroff** [comment]
If I get it right, it expects method `use_relative_model_naming?` defined on some of parent classes and destroys other segments. So if we talk about removing namespace as general behavior, then we can easily use class name, instead of `model_name`.

**Comment by beauby** [comment]
We should split your work into 3 PRs IMO, one for the `type` directive, one for the `type_model` config, and one for removing the namespace.
My point for `model_name` vs `name` is that we should not modify something unless there's a good reason for it, so that we do not risk changing behavior for no reason.

**Comment by youroff** [comment]
Ok, I'll do that soon.
For model name, I think if we remove namespace as a PR, we have to get rid of model_name anyway, because otherwise we'll call it like `object.class.model_name.name.demodulize.underscore` which is definitely longer way to the same point as `object.class.name.demodulize.underscore`

**Comment by beauby** [comment]
Sure, let's include that change in the PR that removes the namespace.

**Comment by beauby** [comment]
Thanks for your work and your implication btw, it is highly appreciated!

**Comment by youroff** [comment]
I have one concern, though. If I remove namespaces in separate commit, it'll be impossible to merge automatically. So namespace removal PR should be made over master, that will already have type_source merged.

**Comment by youroff** [comment]
Just submitted a PR for namespaces. `type_source` one can be safely made only over this one, so this should be merged first to avoid mess. 
https://github.com/rails-api/active_model_serializers/pull/1216

File: lib/active_model/serializer.rb [hunk 9]
```diff
@@ -46,6 +47,10 @@ def self.inherited(base)
       super
     end
 
+    def self.type(type)
+      self._type = type
+    end
```
Comment (bf4) [inline lib/active_model/serializer.rb]: Awesome! :100: Thanks for helping keep the public interface narrow!

Comment (youroff) [inline lib/active_model/serializer.rb]: I'm not sure it's narrow. Underscored class-level accessor will be accessible anyways.

File: test/adapter/json_api/resource_type_config_test.rb [hunk 10]
```diff
@@ -52,6 +52,13 @@ def test_config_singular
               end
             end
           end
+
+          def test_config_custom
```
Comment (beauby) [inline test/adapter/json_api/resource_type_config_test.rb]: Why is it named "config"?

File: lib/active_model/serializer.rb (hunk 9 above)
Comment (bf4) [inline lib/active_model/serializer.rb]: well, basically we want to avoid creating internal instance methods that collide with attribute names.. less important in class variables, but still :)

File: lib/active_model/serializer/adapter/json_api.rb [hunk 11]
```diff
@@ -71,6 +71,7 @@ def serializable_hash_for_single_resource(options)
         end
 
         def resource_identifier_type_for(serializer)
+          return serializer.class._type unless serializer.class._type.nil?
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Might be nice to use `class_attribute :_type, instance_writer: false` so we can just `_type` here

File: test/fixtures/poro.rb [hunk 12]
```diff
@@ -68,6 +68,11 @@ class ProfilePreviewSerializer < ActiveModel::Serializer
   attributes :name
 end
 
+class ProfileTypeSerializer < ActiveModel::Serializer
+  attributes :name
+  type 'profile'
+end
```
Comment (bf4) [inline test/fixtures/poro.rb]: let's define this inside the test.  this file is betting kind of instance

File: test/adapter/json_api/resource_type_config_test.rb (hunk 10 above)
Comment (bf4) [inline test/adapter/json_api/resource_type_config_test.rb]: `test_explicit_type`

File: test/fixtures/poro.rb (hunk 12 above)
Comment (youroff) [inline test/fixtures/poro.rb]: Haha... for me whole this minitest set is already insane. I agree, so I define it inside of file with test?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 11 above)
Comment (youroff) [inline lib/active_model/serializer/adapter/json_api.rb]: Ok

File: test/adapter/json_api/resource_type_config_test.rb (hunk 10 above)
Comment (youroff) [inline test/adapter/json_api/resource_type_config_test.rb]: Just copying name convention from sibling tests, no idea why they have config :)

File: lib/active_model/serializer.rb (hunk 9 above)
Comment (youroff) [inline lib/active_model/serializer.rb]: Yeah, right, I see that instance should be as clean as possible to let users use their names for attributes.

**Comment by youroff** [comment]
Made all corrections mentioned

File: lib/active_model/serializer.rb [hunk 13]
```diff
@@ -37,6 +37,7 @@ class << self
       attr_accessor :_cache_except
       attr_accessor :_cache_options
       attr_accessor :_cache_digest
+      attr_accessor :_type
```
Comment (bf4) [inline lib/active_model/serializer.rb]: I wish there were a better name for this, or a way to signal it is for json_api... `resource_type_name` ? no need to do anything

Comment (bf4) [inline lib/active_model/serializer.rb]: n/m, it can/should be removed since `class_attribute :_type, instance_writer: false` is added below

File: lib/active_model/serializer/adapter/json_api.rb [hunk 14]
```diff
@@ -71,6 +71,7 @@ def serializable_hash_for_single_resource(options)
         end
 
         def resource_identifier_type_for(serializer)
+          return serializer._type unless serializer._type.nil?
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: simpler `return serializer._type if serializer._type` no?

File: lib/active_model/serializer/adapter/json_api.rb [hunk 15]
```diff
@@ -71,6 +71,7 @@ def serializable_hash_for_single_resource(options)
...
           if ActiveModel::Serializer.config.jsonapi_resource_type == :singular
             serializer.object.class.model_name.singular
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: I know you didn't write this, but urgh

Comment (NullVoxPopuli) [inline lib/active_model/serializer/adapter/json_api.rb]: :+1:

File: test/adapter/json_api/resource_type_config_test.rb [hunk 16]
```diff
@@ -35,6 +35,11 @@ def with_jsonapi_resource_type type
             ActiveModel::Serializer.config.jsonapi_resource_type = old_type
           end
 
+          ProfileTypeSerializer = Class.new(ActiveModel::Serializer) do
+            attributes :name
+            type 'profile'
+          end
+
```
Comment (bf4) [inline test/adapter/json_api/resource_type_config_test.rb]: Let's do regular classes plz :)  I don't know what the previous ones were like this.

```ruby
class ResourceTypeConfigTest < Minitest::Test
  class ProfileTypeSerializer < ActiveModel::Serializer
    attributes :name
    type 'profile'
   end
```

File: test/adapter/json_api/resource_type_config_test.rb [hunk 17]
```diff
@@ -52,6 +57,13 @@ def test_config_singular
               end
             end
           end
+
+          def test_explicit_type_value
+            with_adapter :json_api do
+              hash = ActiveModel::SerializableResource.new(@author, serializer: ProfileTypeSerializer).serializable_hash
+              assert_equal('profile', hash[:data][:type])
```
Comment (bf4) [inline test/adapter/json_api/resource_type_config_test.rb]: `with_adapter` is a [global thing that needs to go away](https://github.com/rails-api/active_model_serializers/blob/master/test/support/serialization_testing.rb#L8-L12).  if you don't mind, let's

```ruby
serialized_resource = ActiveModel::SerializableResource.new(@author, serializer: ProfileTypeSerializer, adapter: :json_api).serializable_hash
assert_equal('profile', serialized_resource.fetch(:data).fetch(:type))
```

(the fetch is just to get a better exception if serialized_resource isn't expected)

if you want to [take advantage of a future test simplification](https://github.com/rails-api/active_model_serializers/pull/1050/files#diff-02cd2f48e8583af0b097a44913eb31baR33), add the private method `serializable` to this test

```ruby
def serializable(resource, options = {})
    ActiveModel::SerializableResource.new(resource, options)
  end
```

so then you have `hash = serializable(@author, serializer: ..., adapter: :json_api).serializable_hash`

**Comment by bf4** [comment]
A few small things and this is ready to merge. :100:

File: lib/active_model/serializer/adapter/json_api.rb (hunk 14 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: :+1: @bf

File: test/adapter/json_api/resource_type_config_test.rb (hunk 16 above)
Comment (beauby) [inline test/adapter/json_api/resource_type_config_test.rb]: :+1:

File: lib/active_model/serializer/adapter/json_api.rb (hunk 14 above)
Comment (youroff) [inline lib/active_model/serializer/adapter/json_api.rb]: How about using `.present?` here? As long as we rely on `ActiveSupport` and empty type field doesn't make too much sense... Using variable directly in conditional would work, but I'm trying to avoid such things. Just in case. :)

File: lib/active_model/serializer.rb (hunk 13 above)
Comment (bf4) [inline lib/active_model/serializer.rb]: - [ ] plz remove this accessor as the class_attribute below duplicates the behavior

File: lib/active_model/serializer.rb [hunk 18]
```diff
@@ -122,6 +127,7 @@ def self.get_serializer_for(klass)
     end
 
     attr_accessor :object, :root, :meta, :meta_key, :scope
+    class_attribute :_type, instance_writer: false
```
Comment (bf4) [inline lib/active_model/serializer.rb]: - [ ] this can go up near the other class methods

File: lib/active_model/serializer/adapter/json_api.rb [hunk 19]
```diff
@@ -71,6 +71,7 @@ def serializable_hash_for_single_resource(options)
         end
 
         def resource_identifier_type_for(serializer)
+          return serializer._type if serializer._type.present?
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: Since the default is nil, anything else is user entered.  This basically just guards against an empty string, but since they want that, let'em :)

I'd prefer `return serializer._type if serializer._type`... since it doesn't have to go all the way up the method call stack... https://github.com/rails/rails/blob/4-0-stable/activesupport/lib/active_support/core_ext/object/blank.rb  

but if you really like present?, not a blocker.  I incline to avoid monkeypatching

Comment (youroff) [inline lib/active_model/serializer/adapter/json_api.rb]: I'm not crazy about `present?` at all and totally ok with having just _type. I just find implicit type-conversion confusing and try to avoid that when possible.

**Comment by bf4** [comment]
Looks good, last thing to do is rebase it to squash it down to one commit.  Lemme know if you want any tips on doing that..

**Comment by youroff** [comment]
If you have some short recipe how to do that, would be nice. Otherwise, I'll google. ;)

**Comment by youroff** [comment]
So what I did:
```
git checkout master
git pull
git checkout type_setting
git rebase master
```
As long as nothing changed since I started this PR, nothing happened... Should I merge to master locally or what?

**Comment by bf4** [comment]
This is pretty good
- http://help.exercism.io/git-workflow.html
- A lot more awesome, but a bit more to learn http://rakeroutes.com/blog/deliberate-git/
- What I would do is `git rebase -i 305ec8dc50a16e462943cdffb4ebe346637c61a7^` turn all the `pick` into `f`, except the first, which I'd change to `r` and rewrite the commit message to `Add Serializer 'type' directive to control type field, for use by the JsonApi adapter`, `git fetch master; git rebase origin/master; git push -f`

**Comment by bf4** [comment]
you missed updating your local master to the remote on origin :) see above for how I squash.. or if you want we can pair or I can do it.. up to you

**Comment by youroff** [comment]
Does it look ok? I expected previous commits to disappear, is that ok that they are still there?

**Comment by bf4** [comment]
beautiful!  GitHub still knows about the other commits... they're just not part of the PR. It git, you need to work hard to remove things :)

File: test/adapter/json_api/resource_type_config_test.rb [hunk 20]
```diff
@@ -36,22 +41,29 @@ def with_jsonapi_resource_type type
...
-              end
+            with_jsonapi_resource_type :singular do
+              hash = serializable(@comment, adapter: :json_api).serializable_hash
+              assert_equal('comment', hash[:data][:type])
             end
           end
+
+          def test_explicit_type_value
+            hash = serializable(@author, serializer: ProfileTypeSerializer, adapter: :json_api).serializable_hash
+            assert_equal('profile', hash.fetch(:data).fetch(:type))
```
Comment (beauby) [inline test/adapter/json_api/resource_type_config_test.rb]: Why using `fetch` here instead of brackets?

Comment (bf4) [inline test/adapter/json_api/resource_type_config_test.rb]: Better error on failure

B mobile phone

Comment (beauby) [inline test/adapter/json_api/resource_type_config_test.rb]: Then why not use it [there](https://github.com/rails-api/active_model_serializers/pull/1213/files#diff-2e5487b4844a235b4d646ae346b76499R53)?

Comment (bf4) [inline test/adapter/json_api/resource_type_config_test.rb]: Not at computer now, could be oversight, not important in any case

B mobile phone

Comment (beauby) [inline test/adapter/json_api/resource_type_config_test.rb]: Agreed it's not important, but best to remain consistent.

Comment (joaomdmoura) [inline test/adapter/json_api/resource_type_config_test.rb]: > remain consistent

please
